File: python/Q202_happy_number.py
#
# @lc app=leetcode id=202 lang=python3
#
# [202] Happy Number
#

# @lc code=start
class Solution:
    def isHappy(self, n: int) -> bool:
        table: list[int] = [0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
        result: set[int] = set()

        while n not in result:
            if n == 1:
                return True

            result.add(n)

            digits: list[int] = []

            while n:
                digit: int = n % 10
                n = n // 10

                digits.append(digit)

            n = sum(table[d] for d in digits)

        return False


# @lc code=end

File: rust/src/leetcode/Q202_happy_number.rs
/*
 * @lc app=leetcode id=202 lang=rust
 *
 * [202] Happy Number
 */

// @lc code=start
use std::collections::HashSet;

impl Solution {
    pub fn is_happy(n: i32) -> bool {
        let mut n: i32 = n;
        let mut numbers: HashSet<i32> = HashSet::new();

        while !numbers.contains(&n) {
            if n == 1 {
                return true;
            }
            numbers.insert(n);

            let mut summation: Vec<i32> = Vec::new();
            while n != 0 {
                let digit: i32 = n % 10;
                n /= 10;
                summation.push(digit * digit);
            }

            n = summation.into_iter().sum();
        }

        false
    }
}
// @lc code=end
